js/symb/leafletMap.js

```diff
@@ -0,0 +1,223 @@
+      color: '#000',
+      opacity: 0.85,
+      fillOpacity: 0.55
+   };
+
+   DEFAULT_DRAW_OPTIONS = {
+      polyline: false,
+      control: true,
+      circlemarker: false,
+      marker: false,
+      multiDraw: false,
+      drawColor: this.DEFAULT_SHAPE_OPTIONS 
+   };
+   
+   /* To Hold Reference to Leaflet Map */
+   mapLayer;
+
+   /* Active_Shape Type
+    * type: String represenation of shape 
+    * layer: leaflet layer
+    * center: { lat: float, lng float }
+    */
+   activeShape;
+
+   //List of drawn shapes
+   shapes = [];
+
+   /* Reference Leaflet Feature Group for all drawn items*/
+   drawLayer;
+
+   constructor(map_id, map_options=this.DEFAULT_MAP_OPTIONS) {
+      this.mapLayer = L.map(map_id, map_options);
+
+      const terrainLayer = L.tileLayer('https://{s}.google.com/vt?lyrs=p&x={x}&y={y}&z={z}', {
+         attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
+         subdomains:['mt0','mt1','mt2','mt3'],
+         maxZoom: 20, 
+         tileSize: 256,
+      }).addTo(this.mapLayer);
+
+      const satelliteLayer = L.tileLayer('https://{s}.google.com/vt/lyrs=s&x={x}&y={y}&z={z}', {
+         attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
+         subdomains:['mt1','mt2','mt3'],
+         maxZoom: 20, 
+         tileSize: 256,
+      });
+
+      L.control.layers({
+         "Terrain": terrainLayer,
+         "Satellite": satelliteLayer
+      }).addTo(this.mapLayer);
+
+      L.control.scale().addTo(this.mapLayer);
+   }
+
+   enableDrawing(drawOptions = this.DEFAULT_DRAW_OPTIONS, onDrawChange) {
+
+      var drawnItems = new L.FeatureGroup();
+      this.drawLayer = drawnItems;
+      this.mapLayer.addLayer(drawnItems);
+
+
+      //Jank workaround for leaflet-draw api
+      const setDrawColor = (drawOption) => {
+         if(drawOptions[drawOption] === false)
+            return;
+
+         if(!drawOptions[drawOption]) drawOptions[drawOption] = { };
+
+         if(!drawOptions[drawOption].shapeOptions)
+            drawOptions[drawOption].shapeOptions = drawOptions.drawColor;
+      }
+
+      if(drawOptions.drawColor) {
+         //Setting Styles for Pre Rendered
+         L.Path.mergeOptions(drawOptions.drawColor);
+         //Set Color for All Manual Draws 
+         setDrawColor("polyline");
+         setDrawColor("polygon");
+         setDrawColor("rectangle");
+         setDrawColor("circle");
+      }
+
+      if(drawOptions.control || drawOptions.control === undefined) {
+         var drawControl = new L.Control.Draw({
+            draw: drawOptions,
+            edit: {
+               featureGroup: drawnItems,
+            }
+         });
+
+         this.mapLayer.addControl(drawControl);
+
+         //Event saved edit 
+         this.mapLayer.on('draw:edited', function(e) {
+            ///Some Extra steps to get at the layer
+            const layer = e.layers._layers;
```

Comment:
Could `layers` or `_layers` ever not exist?

Reply:
I don't believe so because that is the under hood loose typing, not sure if they are using JavaScript docs typescript or just painstaking ensuring the layer exists but event should always be tied to an layer